controller/userController.js

```diff
@@ -86,7 +89,174 @@ const token = (req, res) => {
     }
 }
 
+const create = (req, res) => {
+    const user = new User(req.body);
+
+    // Check if an user with the provided email exists
+    User.find({ email: user.email }).then(users => {
+        if (users.length == 0) {
+            // Generate salt & passwordHash
+            user.salt = bcrypt.genSaltSync(AppConfig.SALT_ROUNDS);
+            user.passwordHash = bcrypt.hashSync(user.passwordHash, user.salt);
+            // Add new user
+            user.save().then(() => {
+                user.passwordHash = null;
+                user.salt = null;
+                return res.status(HttpStatus.OK).json(user);
+            });
+        } else {
+            return res.status(HttpStatus.BAD_REQUEST).json({
+                status: 'Error',
+                message: 'User with provided email already exists!'
+            });
+        }
+    }).catch(err => {
+        const errorMessage = ValidationUtil.buildErrorMessage(err, 'create', 'user');
+        return res.status(HttpStatus.BAD_REQUEST).json({
+            status: 'Error',
+            message: errorMessage
+        });
+    });
+};
+
+const retrieveAll = (req, res) => {
+    User.find({}, { passwordHash: 0, salt: 0 }).then(users => {
+        if (users) {
+            return res.status(HttpStatus.OK).json(users);
+        } else {
+            return res.status(HttpStatus.NOT_FOUND).json({
+                status: 'Error',
+                message: 'Users not found.'
+            });
+        }
+    }).catch(err => {
+        const errorMessage = ValidationUtil.buildErrorMessage(err, 'retrieveAll', 'user');
+        return res.status(HttpStatus.INTERNAL_SERVER_ERROR).json({
+            status: 'Error',
+            message: errorMessage
+        });
+    });
+};
+
+const retrieveById = (req, res) => {
+    const userId = req.params.id;
+    if (userId.length != AppConfig.OBJECT_ID_LEN) {
+        return res.status(HttpStatus.BAD_REQUEST).json({
+            status: 'Error',
+            message: 'Invalid ID'
+        });
+    }
+
+    User.findById(userId, { passwordHash: 0, salt: 0 }).then(user => {
+        if (user) {
+            return res.status(HttpStatus.OK).json(user);
+        } else {
+            return res.status(HttpStatus.NOT_FOUND).json({
+                status: 'Error',
+                message: 'User not found.'
+            });
+        }
+    }).catch(err => {
+        const errorMessage = ValidationUtil.buildErrorMessage(err, 'retrieveById', 'user');
+        return res.status(HttpStatus.INTERNAL_SERVER_ERROR).json({
+            status: 'Error',
+            message: errorMessage
+        });
+    });
+};
+
+const update = (req, res) => {
+    const userId = req.params.id;
+    if (userId.length != AppConfig.OBJECT_ID_LEN) {
+        return res.status(HttpStatus.BAD_REQUEST).json({
+            status: 'Error',
+            message: 'Invalid ID'
+        });
+    }
+
+    User.findById(userId).then(user => {
```

Comment:
Mongoose models have a findByIdAndUpdate function, so this could be simplified to use it without needing two calls. 
https://mongoosejs.com/docs/api.html#model_Model.findByIdAndUpdate